src/track/bpm.h

```diff
@@ -83,25 +98,95 @@ class Bpm final {
         return displayValueText(m_value);
     }
 
+    Bpm& operator+=(double increment) {
+        DEBUG_ASSERT(hasValue());
+        m_value += increment;
+        return *this;
+    }
+
+    Bpm& operator-=(double decrement) {
+        DEBUG_ASSERT(hasValue());
+        m_value -= decrement;
+        return *this;
+    }
+
+    Bpm& operator*=(double multiple) {
+        DEBUG_ASSERT(hasValue());
+        m_value *= multiple;
+        return *this;
+    }
+
+    Bpm& operator/=(double divisor) {
+        DEBUG_ASSERT(hasValue());
+        m_value /= divisor;
+        return *this;
+    }
+
 private:
     double m_value;
 };
 
-inline
-bool operator==(const Bpm& lhs, const Bpm& rhs) {
-    return lhs.compareEq(rhs);
+/// Bpm can be added to a double
+inline Bpm operator+(Bpm bpm, double bpmDiff) {
+    return Bpm(bpm.getValue() + bpmDiff);
 }
 
-inline
-bool operator!=(const Bpm& lhs, const Bpm& rhs) {
-    return !(lhs == rhs);
+/// Bpm can be subtracted from a double
+inline Bpm operator-(Bpm bpm, double bpmDiff) {
+    return Bpm(bpm.getValue() - bpmDiff);
 }
 
-inline
-QDebug operator<<(QDebug dbg, const Bpm& arg) {
-    return dbg << arg.getValue();
+/// Two Bpm values can be subtracted to get a double
+inline double operator-(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() - bpm2.getValue();
 }
 
+// Adding two Bpm is not allowed, because it makes no sense semantically.
+
+/// Bpm can be multiplied or divided by a double
+inline Bpm operator*(Bpm bpm, double multiple) {
+    return Bpm(bpm.getValue() * multiple);
+}
+
+inline Bpm operator/(Bpm bpm, double divisor) {
+    return Bpm(bpm.getValue() / divisor);
+}
+
+inline bool operator<(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() < bpm2.getValue();
+}
+
+inline bool operator<=(Bpm bpm1, Bpm bpm2) {
+    return (!bpm1.hasValue() && !bpm2.hasValue()) || bpm1.getValue() <= bpm2.getValue();
+}
+
+inline bool operator>(Bpm bpm1, Bpm bpm2) {
```

Comment:
The definitions of the comparison operators are inconsistent, i.e. the result of `lhs < rhs` should equal `!(lhs >= rhs)`.

Reply:
In which cases does it not?

Reply:
Note that the `<` operator uses `getValue()` which throws a debug assert if at least one of the values is invalid.

100 < 100 = false
100 < invalid = debug assert
invalid < 100 = debug assert
invalid < invalid = debug assert

100 <= 100 = true
100 <= invalid = debug assert
invalid <= 100 = debug assert
invalid <= invalid = true (because they are equal)

Reply:
Independent of debug assertions these operators must be interchangeable. Everything else is unexpected.

The condition `<` must result in the same behavior as `not >=` because as a developer I expect that I could use either of them depending on which is more suitable for the structuring the code.

Reply:
If I see an expression like `if (!(a >= b))` I will refactor it to `if (a < b)`.

Reply:
getValue() always returns 0.0 if the value is invalid, so it should still work fine:

50 < 100 = true
100 < 50 = false
100 < 100 = false
100 < invalid = false + debug assert
invalid < 100 = true + debug assert
invalid < invalid = false debug assert

50 >= 100 = false
100 >= 100 = true
100 >= 100 = true
100 >= invalid = true + debug assert
invalid >= 100 = false + debug assert
invalid >= invalid = true

Reply:
Usually you only implement `<` and then the remaining operators in terms of this to prevent inconsistencies.